### PR TITLE
Map Cassandra INET to Trino IPADDRESS type

### PR DESCRIPTION
**Pull request by chen-ni**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

An improvement in type mapping. We used to map Cassandra `INET` type to Trino `VARCHAR`. It would be better to map it to Trino `IPADDRESS`.

Due to the fact that `IpAddressType` is not located in io.trino.spi.type, we have to use `typeManager().getType()` to get the type reference (please refer to [this discussion](https://trinodb.slack.com/archives/CP1MUNEUX/p1656779968805339?thread_ts=1656754605.965319&cid=CP1MUNEUX)). 

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

The Cassandra connector.

> How would you describe this change to a non-technical end user or system administrator?

By mapping the Cassandra `INET` to Trino `IPADDRESS` type, when reading `INET` data in Cassandra, we can apply [IP Address Functions](https://trino.io/docs/current/functions/ipaddress.html) without the need for casting.

## Related issues, pull requests, and links
Fix https://github.com/trinodb/trino/issues/851.

I've referred to the 2 following PRs:

- https://github.com/trinodb/trino/pull/5231
- https://github.com/trinodb/trino/pull/3347

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Cassandra
* Map Cassandra `inet` type to Trino `ipaddress` type. ({issue}`851`)
```

### CODE REVIEW

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 1]
```diff
@@ -136,6 +137,7 @@ protected QueryRunner createQueryRunner()
         server = closeAfterClass(new CassandraServer());
         session = server.getSession();
         session.execute("CREATE KEYSPACE IF NOT EXISTS " + KEYSPACE + " WITH REPLICATION = {'class':'SimpleStrategy', 'replication_factor': 1}");
+        //return createCassandraQueryRunner(server, ImmutableMap.of(), ImmutableMap.of(), ImmutableList.of());
```
Comment (findinpath) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:140]: pls remove useless code

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java [hunk 2]
```diff
@@ -48,7 +42,7 @@ private CassandraTypes() {}
     public static final CassandraType DECIMAL = new CassandraType(Kind.DECIMAL, DoubleType.DOUBLE);
     public static final CassandraType DOUBLE = new CassandraType(Kind.DOUBLE, DoubleType.DOUBLE);
     public static final CassandraType FLOAT = new CassandraType(Kind.FLOAT, RealType.REAL);
-    public static final CassandraType INET = new CassandraType(Kind.INET, createVarcharType(IP_ADDRESS_STRING_MAX_LENGTH));
+    public static CassandraType INET; // Will be initialized in CassandraMetadata
```
Comment (findinpath) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java:45]: This comment is rather confusing.
Why would we need to do this? Take into account also the fact that multiple requests are handled in the same time or over the lifetime of Trino process, so this field would be initialized over and over multiple times.

**Comment by findinpath** [comment]
Pls remove the first two commits (they are useless if the second one reverts the first one).

**Comment by findinpath** [comment]
Please do take some time to fill in the response about the question:

> How would you describe this change to a non-technical end user or system administrator?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java (hunk 2 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java:45]: @findinpath You're right, I didn't realize it would be initialized over and over again.
I just want to initialize it once, but couldn't find an approach to do it gracefully. I need to call `typeManager().getType()` to get the type reference, but that doesn't seem to be possible in a static context like this..

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 3]
```diff
@@ -1200,6 +1202,39 @@ public void testAllTypesInsert()
...
+                        partitionColumn("key", "text"),
+                        generalColumn("typeinet", "inet")),
+                ImmutableList.of())) {
+            String sql = "SELECT key, typeinet" +
+                    " FROM " + testCassandraTable.getTableName();
+            assertEquals(execute(sql).getRowCount(), 0);
+
+            // TODO Following types are not supported now. We need to change null into the value after fixing it
+            // blob, frozen<set<type>>, inet, list<type>, map<type,type>, set<type>, decimal, varint
+            // timestamp can be inserted but the expected and actual values are not same
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1220]: Remove unhelpful comment.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 4]
```diff
@@ -1200,6 +1202,39 @@ public void testAllTypesInsert()
...
+            execute("INSERT INTO " + testCassandraTable.getTableName() + " (" +
+                    "key," +
+                    "typeinet" +
+                    ") VALUES (" +
+                    "'key1', " +
+                    "IPADDRESS '10.10.10.1'" +
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1226]: Please add more test cases, e.g. min, max, null, ipv6.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/cassandra/TestInsertIntoCassandraTable.java [hunk 5]
```diff
@@ -100,30 +102,35 @@ public void testInsertIntoValuesToCassandraTableAllSimpleTypes()
...
+                            "ascii value",
+                            99999,
+                            null,
+                            true,
+                            null,
+                            123.456789,
+                            Date.valueOf("9999-12-31"),
+                            123.45678,
+                            null,
+                            Inet4Address.getByName("0.0.0.0"),
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/cassandra/TestInsertIntoCassandraTable.java:117]: Use `InetAddresses.forString` and remove try-catch block.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/cassandra/TestInsertIntoCassandraTable.java [hunk 6]
```diff
@@ -100,30 +102,35 @@ public void testInsertIntoValuesToCassandraTableAllSimpleTypes()
...
+                            -32768,
+                            "text value",
+                            -128,
+                            Timestamp.from(OffsetDateTime.of(9999, 12, 31, 23, 59, 59, 0, ZoneOffset.UTC).toInstant()),
+                            "50554d6e-29bb-11e5-b345-feff819cdc9f",
+                            "12151fd2-7586-11e9-8f9e-2a86e4085a59",
+                            "varchar value",
+                            null));
+        }
+        catch (UnknownHostException e) {
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/cassandra/TestInsertIntoCassandraTable.java:131]: Note that just throwing `Exception` is fine in tests.

**Review by ebyhr — COMMENTED** [review]
> I’d like to know if this approach is promising. 

You can refactor `CassandraTypes` first to graduate static type mapping. We wanted to make the mapping dynamic for `LIST`, `MAP` and `SET` type.

**Review by ebyhr — COMMENTED** [review]
Please update "Data types" in `cassandra.rst`. Thanks for your contribution.

**Comment by chen-ni** [comment]
> You can refactor `CassandraTypes` first to graduate static type mapping.

@ebyhr Thanks for the helpful suggestion!

To move away from the static type mapping pattern, the simplest approach I can think of is just to make those fields non-static, initialize them from the constructor (and maybe add some builders later to allow dynamically getting type mappings)  and access those type mappings through an instance of `CassandraTypes`. 

Is this what's expected? Or is some bigger change expected?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 4 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java]: @ebyhr Just found out there is already a test dedicated to testing all the edge cases of the `INET` type: `TestCassandraTypeMapping.testCassandraInet`.

I removed this `TestCassandraConnectorTest.testInetInsert` altogether because it's a temporary test I added to get fast feedback. It was extracted from `TestCassandraConnectorTest.testAllTypesInsert` so it's already covered and was pushed by accident.

**Review by ebyhr — COMMENTED** [review]
> Is this what's expected? Or is some bigger change expected?

I expect larger change. We shouldn't initialize a CassandraTypes variable from CassandraMetadata.

**Comment by chen-ni** [comment]
> We shouldn't initialize a CassandraTypes variable from CassandraMetadata.

@ebyhr I agree. I'm trying to figure out how to avoid it, but it seems a little difficult.

The biggest problem is how to get a `typeManager` instance in the static context of `CassandraTypes`. Perhaps we could consider using [static injection in Guice](https://github.com/google/guice/wiki/Injections#static-injections)? But it's not a recommended approach either.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java (hunk 2 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java]: @findinpath I just removed the static `INET` type altogether. Please check the update if possible, thanks!

**Comment by chen-ni** [comment]
@ebyhr I just removed the static `INET` type in `CassandraTypes` altogether. Please check if this approach is acceptable. Thanks!

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 7]
```diff
@@ -189,7 +196,9 @@ public static Optional<CassandraType> toCassandraType(DataType dataType)
             case ProtocolConstants.DataType.FLOAT:
                 return Optional.of(CassandraTypes.FLOAT);
             case ProtocolConstants.DataType.INET:
-                return Optional.of(CassandraTypes.INET);
+                return Optional.of(new CassandraType(
```
Comment (findinpath) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:199]: Is the new dependecy towards `typeManager` added because `IpAddressType` is only on `trino-main` (and not `trino-spi`) ?

Comment (findinpath) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:199]: Nevermind, addressed in https://trinodb.slack.com/archives/CP1MUNEUX/p1656779968805339?thread_ts=1656754605.965319&cid=CP1MUNEUX

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 8]
```diff
@@ -165,7 +172,7 @@ public String getName()
         return kind.name();
     }
 
-    public static Optional<CassandraType> toCassandraType(DataType dataType)
+    public static Optional<CassandraType> toCassandraType(DataType dataType, TypeManager typeManager)
```
Comment (findinpath) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:175]: Can you please consider extracting the static methods from `CassandraType` to a Guice bean which has the `TypeManager typeManager` as `private final` field ?

This would leave `CassandraType` a simple POJO and the newly created bean could get eventually further dependencies, but further expose a simple public interface to the client classes.

Comment (ebyhr) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:175]: @chen-ni You can refer to https://github.com/ebyhr/trino/commit/10742d16940f9b970e0b05bce37d766d35f5d8f4 as the example. Please note that extracting those methods in a commit may make git diff dirty. You may need additional commits for keeping clean diff.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java (hunk 7 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:199]: @findinpath Yes, I should have added the link in the PR description (I'll add now), sorry for the confusion.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 9]
```diff
@@ -256,7 +265,7 @@ private static Optional<CassandraType> createTypeForUserType(DataType dataType)
             throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Mismatch between the number of field names (%s) and the number of field types (%s) for the data type %s", fieldNames.size(), fieldTypes.size(), dataType));
         }
         for (int i = 0; i < fieldNames.size(); i++) {
-            Optional<CassandraType> cassandraType = CassandraType.toCassandraType(fieldTypes.get(i));
```
[the remark this answers is not in the document]
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:259]: @findinpath Sorry, I don't quite understand here. What do we need to put in the separate commit? A change to static import?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java (hunk 8 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:175]: @findinpath @ebyhr Thanks, I'll give it a try  :)

Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java]: @ebyhr I just removed my old commit, cherry-picked your refactoring commit, then put the `INET -> IPADDRESS` type mapping part based on it. So now we have 2 commits, 1 for the refactoring and 1 for the type mapping. Hopefully it will work.

Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java]: @ebyhr @findinpath Integration with @ebyhr 's refactoring looks fine. Please review when possible :)

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 10]
```diff
@@ -91,81 +95,16 @@
 
 public class CassandraType
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:102]: Let's rename to `CassandraTypeManager` in the 1st commit.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 11]
```diff
@@ -91,81 +95,16 @@
...
-        this.argumentTypes = ImmutableList.copyOf(requireNonNull(argumentTypes, "argumentTypes is null"));
-    }
-
-    @JsonProperty
-    public Kind getKind()
-    {
-        return kind;
-    }
+    @SuppressWarnings("unused")
+    private final TypeManager typeManager;
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:104]: Please remove `TypeManager` from the 1st commit (I added this to my example just for showcase).

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java [hunk 12]
```diff
@@ -61,7 +63,7 @@ public void testBuildClusteringPredicate()
                         col1, Domain.singleValue(BIGINT, 23L),
                         col2, Domain.singleValue(BIGINT, 34L),
                         col4, Domain.singleValue(BIGINT, 26L)));
-        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
+        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(new CassandraType(TESTING_TYPE_MANAGER), cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java:66]: Replace `new CassandraType(TESTING_TYPE_MANAGER)` with `CASSANDRA_TYPE_MANAGER`

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java [hunk 13]
```diff
@@ -73,7 +75,7 @@ public void testGetUnenforcedPredicates()
                 ImmutableMap.of(
                         col2, Domain.singleValue(BIGINT, 34L),
                         col4, Domain.singleValue(BIGINT, 26L)));
-        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
+        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(new CassandraType(TESTING_TYPE_MANAGER), cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java:78]: Replace `new CassandraType(TESTING_TYPE_MANAGER)` with `CASSANDRA_TYPE_MANAGER`

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 14]
```diff
@@ -91,11 +96,11 @@
 import static java.lang.Float.intBitsToFloat;
 import static java.lang.Math.toIntExact;
 import static java.lang.String.format;
+import static java.lang.System.arraycopy;
 import static java.util.Objects.requireNonNull;
 
 public class CassandraType
 {
-    @SuppressWarnings("unused")
     private final TypeManager typeManager;
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:104]: Please define `Type ipAddressType` and initialize it in the constructor. Then, remove this `typeManager`.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java [hunk 15]
```diff
@@ -652,6 +664,39 @@ public CassandraTypeMapping toCassandraType(Type type, ProtocolVersion protocolV
...
+        if (type.equals(typeManager.getType(new TypeSignature(StandardTypes.IPADDRESS)))) {
+            return new CassandraTypeMapping(
+                    CassandraTypeMapping.Kind.INET,
+                    typeManager.getType(new TypeSignature(StandardTypes.IPADDRESS)));
+        }
         throw new TrinoException(NOT_SUPPORTED, "Unsupported type: " + type);
     }
+
+    // This is a copy of IpAddressOperators.castFromVarcharToIpAddress method
+    public static Slice castFromVarcharToIpAddress(Slice slice)
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraType.java:676]: nit: Change to private.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeMapping.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeMapping.java:1]: Please add another commit to rename `CassandraTypeMapping` to `CassandraType` as the 2nd commit.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraPageSink.java [hunk 17]
```diff
@@ -185,6 +192,9 @@ else if (VARBINARY.equals(type)) {
         else if (UuidType.UUID.equals(type)) {
             values.add(trinoUuidToJavaUuid(type.getSlice(block, position)));
         }
+        else if (typeManager.getType(new TypeSignature(StandardTypes.IPADDRESS)).equals(type)) {
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraPageSink.java:195]: Please consider adding `boolean isIpAddressType(Type type)` to `CassandraTypeManager` and use here.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java [hunk 18]
```diff
@@ -107,11 +105,17 @@
...
     private final Supplier<CqlSession> session;
     private final Duration noHostAvailableRetryTimeout;
+    private final CassandraTypeManager cassandraTypeManager;
 
-    public CassandraSession(JsonCodec<List<ExtraColumnMetadata>> extraColumnMetadataCodec, Supplier<CqlSession> sessionSupplier, Duration noHostAvailableRetryTimeout)
+    public CassandraSession(
+            JsonCodec<List<ExtraColumnMetadata>> extraColumnMetadataCodec,
+            Supplier<CqlSession> sessionSupplier,
+            Duration noHostAvailableRetryTimeout,
+            CassandraTypeManager cassandraTypeManager)
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:114]: Let's reorder `CassandraTypeManager` to the 1st argument (including `private final...` & `this.cassandraTypeManager...`). Sorry, I know it's comes from my rough example.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java [hunk 19]
```diff
@@ -11,6 +11,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java:14]: Revert unrelated change.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java [hunk 20]
```diff
@@ -61,7 +62,7 @@ public void testBuildClusteringPredicate()
                         col1, Domain.singleValue(BIGINT, 23L),
                         col2, Domain.singleValue(BIGINT, 34L),
                         col4, Domain.singleValue(BIGINT, 26L)));
-        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
+        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(new CassandraTypeManager(), cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java]: nit: Replace `new CassandraTypeManager()` with `CASSANDRA_TYPE_MANAGER`.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java [hunk 21]
```diff
@@ -73,7 +74,7 @@ public void testGetUnenforcedPredicates()
                 ImmutableMap.of(
                         col2, Domain.singleValue(BIGINT, 34L),
                         col4, Domain.singleValue(BIGINT, 26L)));
-        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
+        CassandraClusteringPredicatesExtractor predicatesExtractor = new CassandraClusteringPredicatesExtractor(new CassandraTypeManager(), cassandraTable.getClusteringKeyColumns(), tupleDomain, cassandraVersion);
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java]: nit: Replace `new CassandraTypeManager()` with `CASSANDRA_TYPE_MANAGER`.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraPageSinkProvider.java [hunk 22]
```diff
@@ -28,12 +29,18 @@
...
 {
+    private final CassandraTypeManager cassandraTypeManager;
     private final CassandraSession cassandraSession;
     private final int batchSize;
 
     @Inject
-    public CassandraPageSinkProvider(CassandraSession cassandraSession, CassandraClientConfig cassandraClientConfig)
+    public CassandraPageSinkProvider(
+            CassandraTypeManager cassandraTypeManager,
+            TypeManager typeManager,
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraPageSinkProvider.java:39]: Remove unused variable.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java [hunk 23]
```diff
@@ -0,0 +1,705 @@
...
+import static java.lang.System.arraycopy;
+
+public class CassandraTypeManager
+{
+    private final Type ipAddressType;
+
+    @Inject
+    public CassandraTypeManager(TypeManager typeManager)
+    {
+        this.ipAddressType = typeManager.getType(new TypeSignature(StandardTypes.IPADDRESS));
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java:107]: Please add `requireNonNull`
```java
requireNonNull(typeManager, "typeManager is null");
```

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraTypeMapping.java [hunk 24]
```diff
@@ -392,14 +392,14 @@ public void testCassandraMap()
...
-                .addRoundTrip("inet", "'ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255'", createVarcharType(45), "CAST('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff' AS varchar(45))")
+                .addRoundTrip("inet", "NULL", IPADDRESS, "CAST(NULL AS ipaddress)")
+                .addRoundTrip("inet", "'0.0.0.0'", IPADDRESS, "CAST('0.0.0.0' AS ipaddress)")
+                .addRoundTrip("inet", "'116.253.40.133'", IPADDRESS, "CAST('116.253.40.133' AS ipaddress)")
+                .addRoundTrip("inet", "'255.255.255.255'", IPADDRESS, "CAST('255.255.255.255' AS ipaddress)")
+                .addRoundTrip("inet", "'::'", IPADDRESS, "CAST('::' AS ipaddress)")
+                .addRoundTrip("inet", "'2001:44c8:129:2632:33:0:252:2'", IPADDRESS, "CAST('2001:44c8:129:2632:33:0:252:2' AS ipaddress)")
+                .addRoundTrip("inet", "'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff'", IPADDRESS, "CAST('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff' AS ipaddress)")
+                .addRoundTrip("inet", "'ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255'", IPADDRESS, "CAST('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff' AS ipaddress)")
```
Comment (ebyhr) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraTypeMapping.java:402]: Now we can create and insert into `IPADDRESS` type in the connector. Let's add another test method.
```java
    @Test
    public void testIpAddress()
    {
        SqlDataTypeTest.create()
                .addRoundTrip("ipaddress", "NULL", IPADDRESS, "CAST(NULL AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress '0.0.0.0'", IPADDRESS, "CAST('0.0.0.0' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress '116.253.40.133'", IPADDRESS, "CAST('116.253.40.133' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress '255.255.255.255'", IPADDRESS, "CAST('255.255.255.255' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress '::'", IPADDRESS, "CAST('::' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress '2001:44c8:129:2632:33:0:252:2'", IPADDRESS, "CAST('2001:44c8:129:2632:33:0:252:2' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff'", IPADDRESS, "CAST('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff' AS ipaddress)")
                .addRoundTrip("ipaddress", "ipaddress 'ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255'", IPADDRESS, "CAST('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff' AS ipaddress)")
                .execute(getQueryRunner(), trinoCreateAndInsert("test_ipaddress"))
                .execute(getQueryRunner(), trinoCreateAsSelect("test_ipaddress"));
    }
```

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java (hunk 20 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/util/TestCassandraClusteringPredicatesExtractor.java]: I made the mistake of replacing it in the 3rd commit while I should have replaced it in the 1st commit.. Just fixed :)

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java (hunk 19 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java:14]: Sorry, didn't quite understand this one.. What unrelated change?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraTypeMapping.java (hunk 24 above)
Comment (chen-ni) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraTypeMapping.java:402]: Added :)

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java (hunk 19 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java:14]: This empty line didn't exist before this commit.

Comment (chen-ni) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypeManager.java]: @ebyhr Got it, removed :)

**Comment by ebyhr** [comment]
Merged, thanks!

**Comment by chen-ni** [comment]
@ebyhr Sorry, just realized I didn't remove the outdated static `INET` type here: https://github.com/trinodb/trino/blob/f8be32e63708808e99a0190573fbbae6e57d7626/plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTypes.java#L51

It was indeed removed in the original commit but the change must have been lost while resolving a conflict during one of the fixes later.

Should I create a simple PR to remove it or it's fine to leave it like this for a while?

P.S. Thanks for the helpful code reviews and refactoring example :)

**Comment by ebyhr** [comment]
@chen-ni You can send another PR. Thanks for finding that.
